cmux.go

```diff
@@ -204,6 +210,27 @@ func (m *cMux) serve(c net.Conn, donec <-chan struct{}, wg *sync.WaitGroup) {
 	}
 }
 
+func (m *cMux) Close() {
+	m.closeDoneChans()
+}
+
+func (m *cMux) closeDoneChans() {
+	select {
+	case <-m.donec:
+		// Already closed. Don't close again
+	default:
+		close(m.donec)
```

Comment:
I know the commit is old, and likely very very stale.

But for anyone looking into making this sort of only-once-close on channels, the whole select block needs to be done under a mutex. Otherwise, two simultaneous calls to this could both reach the `default:` case at the same time, and attempt to close the channel twice.